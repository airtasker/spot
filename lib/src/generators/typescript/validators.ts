import * as ts from "typescript";
import assertNever from "../../assert-never";
import {
  Api,
  ArrayType,
  ObjectType,
  OptionalType,
  Type,
  TypeReference,
  UnionType
} from "../../models";
import { outputTypeScriptSource } from "./ts-writer";
import { typeNode } from "./types";
import flatten = require("lodash/flatten");
import compact = require("lodash/compact");

export function generateValidatorsSource(api: Api): string {
  const typeNames = Object.keys(api.types);
  return outputTypeScriptSource([
    ...(typeNames.length > 0
      ? [
          ts.createImportDeclaration(
            /*decorators*/ undefined,
            /*modifiers*/ undefined,
            ts.createImportClause(
              /*name*/ undefined,
              ts.createNamedImports(
                typeNames.map(typeName =>
                  ts.createImportSpecifier(
                    /*propertyName*/ undefined,
                    ts.createIdentifier(typeName)
                  )
                )
              )
            ),
            ts.createStringLiteral("./types")
          )
        ]
      : []),
    ...flatten(
      Object.entries(api.endpoints).map(([endpointName, endpoint]) => [
        generateValidator(
          endpointPropertyTypeName(endpointName, "request"),
          endpoint.requestType
        ),
        ...compact(
          endpoint.path.map(
            pathComponent =>
              pathComponent.kind === "dynamic"
                ? generateValidator(
                    endpointPropertyTypeName(
                      endpointName,
                      "param",
                      pathComponent.name
                    ),
                    pathComponent.type
                  )
                : null
          )
        ),
        ...Object.entries(endpoint.headers).map(([headerName, header]) =>
          generateValidator(
            endpointPropertyTypeName(endpointName, "header", headerName),
            header.type
          )
        ),
        generateValidator(
          endpointPropertyTypeName(endpointName, "response"),
          endpoint.responseType
        ),
        generateValidator(
          endpointPropertyTypeName(endpointName, "genericError"),
          endpoint.genericErrorType
        ),
        ...Object.entries(endpoint.specificErrorTypes).map(
          ([name, specificError]) =>
            generateValidator(
              endpointPropertyTypeName(endpointName, `specificError`, name),
              specificError.type
            )
        )
      ])
    ),
    ...Object.entries(api.types).map(([typeName, type]) =>
      generateValidator(typeName, type, typeName)
    )
  ]);
}

function generateValidator(
  typeNickname: string,
  type: Type,
  typeName?: string
): ts.Statement {
  const parameterName = ts.createIdentifier("value");
  return ts.createFunctionDeclaration(
    /*decorators*/ undefined,
    /*modifiers*/ [ts.createToken(ts.SyntaxKind.ExportKeyword)],
    /*asteriskToken*/ undefined,
    ts.createIdentifier(validatorName(typeNickname)),
    /*typeParameters*/ undefined,
    [
      ts.createParameter(
        /*decorators*/ undefined,
        /*modifiers*/ undefined,
        /*dotDotDotToken*/ undefined,
        parameterName,
        /*questionToken*/ undefined,
        ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
      )
    ],
    /*returnType*/ ts.createTypePredicateNode(
      parameterName,
      typeName
        ? ts.createTypeReferenceNode(typeName, /*typeArguments*/ undefined)
        : typeNode(type)
    ),
    ts.createBlock(
      [ts.createReturn(validator(type, parameterName))],
      /*multiline*/ true
    )
  );
}

function validator(type: Type, parameter: ts.Expression): ts.Expression {
  switch (type.kind) {
    case "void":
      return voidValidator(parameter);
    case "null":
      return nullValidator(parameter);
    case "boolean":
      return booleanValidator(parameter);
    case "boolean-constant":
      return booleanConstantValidator(type.value, parameter);
    case "date":
    case "date-time":
    case "string":
      return stringValidator(parameter);
    case "string-constant":
      return stringConstantValidator(type.value, parameter);
    case "int32":
    case "int64":
    case "float":
    case "double":
    case "number":
      return numberValidator(parameter);
    case "integer-constant":
      return integerConstantValidator(type.value, parameter);
    case "object":
      return objectValidator(type, parameter);
    case "array":
      return arrayValidator(type, parameter);
    case "optional":
      return optionalValidator(type, parameter);
    case "union":
      return unionValidator(type, parameter);
    case "type-reference":
      return typeReferenceValidator(type, parameter);
    default:
      throw assertNever(type);
  }
}

function voidValidator(parameter: ts.Expression): ts.Expression {
  return ts.createStrictEquality(parameter, ts.createIdentifier("undefined"));
}

function nullValidator(parameter: ts.Expression): ts.Expression {
  return ts.createStrictEquality(parameter, ts.createIdentifier("null"));
}

function booleanValidator(parameter: ts.Expression): ts.Expression {
  return ts.createStrictEquality(
    ts.createTypeOf(parameter),
    ts.createStringLiteral("boolean")
  );
}

function stringValidator(parameter: ts.Expression): ts.Expression {
  return ts.createStrictEquality(
    ts.createTypeOf(parameter),
    ts.createStringLiteral("string")
  );
}

function numberValidator(parameter: ts.Expression): ts.Expression {
  return ts.createStrictEquality(
    ts.createTypeOf(parameter),
    ts.createStringLiteral("number")
  );
}

function notNullObjectValidator(parameter: ts.Expression): ts.Expression {
  return ts.createLogicalAnd(
    ts.createLogicalNot(nullValidator(parameter)),
    ts.createStrictEquality(
      ts.createTypeOf(parameter),
      ts.createStringLiteral("object")
    )
  );
}

function booleanConstantValidator(
  value: boolean,
  parameter: ts.Expression
): ts.Expression {
  return ts.createStrictEquality(
    parameter,
    ts.createIdentifier(value ? "true" : "false")
  );
}

function stringConstantValidator(
  value: string,
  parameter: ts.Expression
): ts.Expression {
  return ts.createStrictEquality(parameter, ts.createStringLiteral(value));
}

function integerConstantValidator(
  value: number,
  parameter: ts.Expression
): ts.Expression {
  return ts.createStrictEquality(
    parameter,
    ts.createNumericLiteral(value.toString(10))
  );
}

function objectValidator(
  type: ObjectType,
  parameter: ts.Expression
): ts.Expression {
  let expression = notNullObjectValidator(parameter);
  for (const [propertyName, propertyType] of Object.entries(type.properties)) {
    expression = ts.createLogicalAnd(
      expression,
      validator(
        propertyType,
        ts.createElementAccess(parameter, ts.createStringLiteral(propertyName))
      )
    );
  }
  return expression;
}

function arrayValidator(
  type: ArrayType,
  parameter: ts.Expression
): ts.Expression {
  const accArgument = ts.createIdentifier("acc");
  const currArgument = ts.createIdentifier("curr");
  return ts.createLogicalAnd(
    ts.createBinary(
      parameter,
      ts.SyntaxKind.InstanceOfKeyword,
      ts.createIdentifier("Array")
    ),
    ts.createCall(
      ts.createPropertyAccess(parameter, "reduce"),
      /*typeArguments*/ undefined,
      [
        ts.createArrowFunction(
          /*modifiers*/ undefined,
          /*typeParameters*/ undefined,
          [
            ts.createParameter(
              /*decorators*/ undefined,
              /*modifiers*/ undefined,
              /*dotDotDotToken*/ undefined,
              accArgument
            ),
            ts.createParameter(
              /*decorators*/ undefined,
              /*modifiers*/ undefined,
              /*dotDotDotToken*/ undefined,
              currArgument
            )
          ],
          /*type*/ undefined,
          /*equalsGreaterThanToken*/ undefined,
          ts.createLogicalAnd(
            accArgument,
            validator(type.elements, currArgument)
          )
        ),
        ts.createIdentifier("true")
      ]
    )
  );
}

function optionalValidator(
  type: OptionalType,
  parameter: ts.Expression
): ts.Expression {
  return ts.createLogicalOr(
    voidValidator(parameter),
    validator(type.optional, parameter)
  );
}

function unionValidator(
  type: UnionType,
  parameter: ts.Expression
): ts.Expression {
  let expression = validator(type.types[0], parameter);
  for (let i = 1; i < type.types.length; i++) {
    expression = ts.createLogicalOr(
      expression,
      validator(type.types[i], parameter)
    );
  }
  return expression;
}

function typeReferenceValidator(type: TypeReference, parameter: ts.Expression) {
  return ts.createCall(
    ts.createIdentifier(validatorName(type.typeName)),
    /*typeArguments*/ undefined,
    [parameter]
  );
}

export function endpointPropertyTypeName(
  endpointName: string,
  property:
    | "request"
    | "param"
    | "header"
    | "response"
    | "genericError"
    | "specificError",
  suffix = ""
) {
  switch (property) {
    case "param":
    case "header":
    case "specificError":
      if (!suffix) {
        throw new Error(`Unexpected ${property} without a suffix`);
      }
      break;
    default:
      // No suffix required.
      if (suffix) {
        throw new Error(`Unexpected ${property} with a suffix`);
      }
  }
  return `${endpointName}_${property}${capitaliseFirstLetter(suffix)}`;
}

export function validatorName(typeName: string) {
  return `validate${capitaliseFirstLetter(typeName)}`;
}

function capitaliseFirstLetter(name: string) {
  if (name.length === 0) {
    return name;
  }
  return `${name[0].toUpperCase()}${name.substr(1)}`;
}

export function validateStatement(
  variable: ts.Expression,
  validatorName: string,
  errorMessage: string
): ts.Statement {
  return ts.createIf(
    ts.createLogicalNot(
      ts.createCall(
        ts.createPropertyAccess(
          ts.createIdentifier("validators"),
          validatorName
        ),
        /*typeArguments*/ undefined,
        [variable]
      )
    ),
    ts.createBlock(
      [
        ts.createThrow(
          ts.createNew(
            ts.createIdentifier("Error"),
            /*typeArguments*/ undefined,
            [
              ts.createTemplateExpression(
                ts.createTemplateHead(`${errorMessage}: `),
                [
                  ts.createTemplateSpan(
                    ts.createCall(
                      ts.createPropertyAccess(
                        ts.createIdentifier("JSON"),
                        "stringify"
                      ),
                      /*typeArguments*/ undefined,
                      [
                        variable,
                        ts.createIdentifier("null"),
                        ts.createNumericLiteral("2")
                      ]
                    ),
                    ts.createTemplateTail("")
                  )
                ]
              )
            ]
          )
        )
      ],
      /*multiLine*/ true
    )
  );
}
