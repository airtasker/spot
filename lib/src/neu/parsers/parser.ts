import * as path from "path";
import { CompilerOptions, Project, SourceFile, ts } from "ts-morph";
import { ApiConfig } from "../../syntax/api";
import { Contract, Endpoint, Security } from "../definitions";
import { LociTable } from "../locations";
import { TypeTable } from "../types";
import { parseEndpoint } from "./endpoint-parser";
import {
  getClassWithDecoratorOrThrow,
  getDecoratorConfigOrThrow,
  getJsDoc,
  getObjLiteralPropOrThrow,
  getPropertyWithDecorator,
  getPropValueAsStringOrThrow,
  getSelfAndLocalDependencies
} from "./parser-helpers";
import { parseSecurityHeader } from "./security-header-parser";

export function parse(sourcePath: string): Contract {
  const project = createProject();

  // Add all dependent files that the project requires
  const sourceFile = project.addExistingSourceFile(sourcePath);
  project.resolveSourceFileDependencies();

  // Validate that the project has no TypeScript syntax errors
  validateProject(project);

  return parseRootSourceFile(sourceFile);
}

/**
 * Parse a root source file to return a contract.
 */
function parseRootSourceFile(file: SourceFile): Contract {
  const lociTable = new LociTable();
  const typeTable = new TypeTable();

  const klass = getClassWithDecoratorOrThrow(file, "api");
  const decorator = klass.getDecoratorOrThrow("api");
  const decoratorConfig = getDecoratorConfigOrThrow(decorator);
  const nameProp = getObjLiteralPropOrThrow<ApiConfig>(decoratorConfig, "name");
  const nameLiteral = getPropValueAsStringOrThrow(nameProp);
  const descriptionDoc = getJsDoc(klass);

  const securityHeaderProp = getPropertyWithDecorator(klass, "securityHeader");

  const security: Security = {
    header:
      securityHeaderProp &&
      parseSecurityHeader(securityHeaderProp, typeTable, lociTable)
  };

  // Add location data
  lociTable.addMorphNode(LociTable.apiClassKey(), klass);
  lociTable.addMorphNode(LociTable.apiDecoratorKey(), decorator);
  lociTable.addMorphNode(LociTable.apiNameKey(), nameProp);
  if (descriptionDoc) {
    lociTable.addMorphNode(LociTable.apiDescriptionKey(), descriptionDoc);
  }

  // Resolve all related files
  const projectFiles = getSelfAndLocalDependencies(file);

  // Parse all endpoints
  const endpoints = projectFiles
    .reduce<Endpoint[]>(
      (acc, currentFile) =>
        acc.concat(
          currentFile
            .getClasses()
            .filter(k => k.getDecorator("endpoint") !== undefined)
            .map(k => parseEndpoint(k, typeTable, lociTable))
        ),
      []
    )
    .sort((a, b) => {
      if (a.name === b.name) {
        throw new Error(`Duplicate endpoint detected: ${a.name}`);
      }
      return b.name > a.name ? -1 : 1;
    });

  return {
    name: nameLiteral.getLiteralText(),
    description: descriptionDoc && descriptionDoc.getComment(),
    types: typeTable.toArray(),
    security,
    endpoints
  };
}

/**
 * Create a new project configured for Spot
 */
function createProject(): Project {
  const compilerOptions: CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.CommonJS,
    strict: true,
    noImplicitAny: true,
    strictNullChecks: true,
    strictFunctionTypes: true,
    strictPropertyInitialization: true,
    noImplicitThis: true,
    alwaysStrict: true,
    noImplicitReturns: true,
    noFallthroughCasesInSwitch: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true,
    baseUrl: "./",
    paths: {
      "@airtasker/spot": [path.join(__dirname, "../lib")]
    }
  };

  // Creates a new typescript program in memory
  return new Project({ compilerOptions });
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
function validateProject(project: Project) {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics
        .map(diagnostic => {
          const message = diagnostic.getMessageText();
          return typeof message === "string"
            ? message
            : message.getMessageText();
        })
        .join("\n")
    );
  }
}
