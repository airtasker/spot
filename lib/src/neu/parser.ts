import * as path from "path";
import { CompilerOptions, Project, ts } from "ts-morph";
import { Contract } from "./definitions";
import { parseContract } from "./parsers/contract-parser";

export function parse(sourcePath: string): Contract {
  const project = createProject();

  // Add all dependent files that the project requires
  const sourceFile = project.addExistingSourceFile(sourcePath);
  project.resolveSourceFileDependencies();

  // Validate that the project has no TypeScript syntax errors
  validateProject(project);

  const result = parseContract(sourceFile);

  if (result.isErr()) throw result.unwrapErr();

  return result.unwrap().contract;
}

/**
 * Create a new project configured for Spot
 */
function createProject(): Project {
  const compilerOptions: CompilerOptions = {
    target: ts.ScriptTarget.ESNext,
    module: ts.ModuleKind.CommonJS,
    strict: true,
    noImplicitAny: true,
    strictNullChecks: true,
    strictFunctionTypes: true,
    strictPropertyInitialization: true,
    noImplicitThis: true,
    alwaysStrict: true,
    noImplicitReturns: true,
    noFallthroughCasesInSwitch: true,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    experimentalDecorators: true,
    baseUrl: "./",
    paths: {
      "@airtasker/spot": [path.join(__dirname, "../lib")]
    }
  };

  // Creates a new typescript program in memory
  return new Project({ compilerOptions });
}

/**
 * Validate an AST project's correctness.
 *
 * @param project an AST project
 */
function validateProject(project: Project) {
  const diagnostics = project.getPreEmitDiagnostics();
  if (diagnostics.length > 0) {
    throw new Error(
      diagnostics
        .map(diagnostic => {
          const message = diagnostic.getMessageText();
          return typeof message === "string"
            ? message
            : message.getMessageText();
        })
        .join("\n")
    );
  }
}
