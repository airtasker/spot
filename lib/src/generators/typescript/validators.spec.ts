import * as fs from "fs-extra";
import * as path from "path";
import {
  arrayType,
  BOOLEAN,
  booleanConstant,
  integerConstant,
  NULL,
  NUMBER,
  objectType,
  optionalType,
  STRING,
  stringConstant,
  typeReference,
  unionType,
  VOID
} from "../../models";
import { parsePath } from "../../parser";
import { generateValidatorsSource } from "./validators";

const EXAMPLES_DIR = path.join(
  __dirname,
  "..",
  "..",
  "..",
  "..",
  "examples",
  "src"
);

describe("TypeScript validators generator", () => {
  describe("produces valid code", () => {
    for (const testCaseName of fs.readdirSync(EXAMPLES_DIR)) {
      if (!fs.lstatSync(path.join(EXAMPLES_DIR, testCaseName)).isDirectory()) {
        continue;
      }
      test(testCaseName, async () => {
        const api = await parsePath(
          path.join(EXAMPLES_DIR, testCaseName, "api.ts")
        );
        const source = generateValidatorsSource(api);
        expect(source).toMatchSnapshot();
      });
    }
  });

  it("generates types for endpoints", () => {
    expect(
      generateValidatorsSource({
        endpoints: {
          example: {
            method: "POST",
            path: [],
            headers: {},
            queryParams: [],
            requestType: VOID,
            responseType: VOID,
            genericErrorType: VOID,
            specificErrorTypes: {
              forbidden: {
                statusCode: 403,
                type: VOID
              },
              notFound: {
                statusCode: 404,
                type: VOID
              }
            }
          }
        },
        types: {}
      })
    ).toMatchInlineSnapshot(`
"export function validateExample_request(value: any): value is void {
    return value === undefined;
}

export function validateExample_response(value: any): value is void {
    return value === undefined;
}

export function validateExample_genericError(value: any): value is void {
    return value === undefined;
}

export function validateExample_specificErrorForbidden(value: any): value is void {
    return value === undefined;
}

export function validateExample_specificErrorNotFound(value: any): value is void {
    return value === undefined;
}"
`);
  });

  describe("generates type validator", () => {
    test("void", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: VOID
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return value === undefined;
}"
`);
    });

    test("null", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: NULL
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return value === null;
}"
`);
    });

    test("boolean", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: BOOLEAN
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return typeof value === \\"boolean\\";
}"
`);
    });

    test("boolean constant", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: booleanConstant(true)
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return value === true;
}"
`);
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: booleanConstant(false)
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return value === false;
}"
`);
    });

    test("string", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: STRING
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return typeof value === \\"string\\";
}"
`);
    });

    test("string constant", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: stringConstant("some constant")
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return value === \\"some constant\\";
}"
`);
    });

    test("number", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: NUMBER
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return typeof value === \\"number\\";
}"
`);
    });

    test("integer constant", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: integerConstant(0)
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return value === 0;
}"
`);
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: integerConstant(123)
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return value === 123;
}"
`);
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: integerConstant(-1000)
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return value === -1000;
}"
`);
    });

    test("object", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: objectType({})
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return !(value === null) && typeof value === \\"object\\";
}"
`);
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: objectType({
              singleField: NUMBER
            })
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"singleField\\"] === \\"number\\";
}"
`);
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: objectType({
              field1: NUMBER,
              field2: STRING,
              field3: BOOLEAN
            })
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"field1\\"] === \\"number\\" && typeof value[\\"field2\\"] === \\"string\\" && typeof value[\\"field3\\"] === \\"boolean\\";
}"
`);
    });

    test("array", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: arrayType(STRING)
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return value instanceof Array && value.reduce((acc, curr) => acc && typeof curr === \\"string\\", true);
}"
`);
    });

    test("optional", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: optionalType(STRING)
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return value === undefined || typeof value === \\"string\\";
}"
`);
    });

    test("union", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: unionType(STRING, NUMBER, BOOLEAN)
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return typeof value === \\"string\\" || typeof value === \\"number\\" || typeof value === \\"boolean\\";
}"
`);
    });

    test("type reference", () => {
      expect(
        generateValidatorsSource({
          endpoints: {},
          types: {
            Example: typeReference("OtherType")
          }
        })
      ).toMatchInlineSnapshot(`
"import { Example } from \\"./types\\";

export function validateExample(value: any): value is Example {
    return validateOtherType(value);
}"
`);
    });
  });
});
