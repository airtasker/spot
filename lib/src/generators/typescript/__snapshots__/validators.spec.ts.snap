// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript validators generator produces valid code multi-file 1`] = `
"import { CreateUserRequest, CreateUserResponse, GetUserResponse } from \\"./types\\";

export function validateCreateUser_request(value: any): value is CreateUserRequest {
    return validateCreateUserRequest(value);
}

export function validateCreateUser_headerAuthToken(value: any): value is string | void {
    return value === undefined || typeof value === \\"string\\";
}

export function validateCreateUser_response(value: any): value is CreateUserResponse {
    return validateCreateUserResponse(value);
}

export function validateCreateUser_genericError(value: any): value is void {
    return value === undefined;
}

export function validateDeleteUser_request(value: any): value is void {
    return value === undefined;
}

export function validateDeleteUser_paramUserId(value: any): value is string {
    return typeof value === \\"string\\";
}

export function validateDeleteUser_headerAuthToken(value: any): value is string {
    return typeof value === \\"string\\";
}

export function validateDeleteUser_response(value: any): value is null {
    return value === null;
}

export function validateDeleteUser_genericError(value: any): value is {
    message: string;
} {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"message\\"] === \\"string\\";
}

export function validateDeleteUser_specificErrorForbidden(value: any): value is {
    message: string;
    signedInAs: string;
} {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"message\\"] === \\"string\\" && typeof value[\\"signedInAs\\"] === \\"string\\";
}

export function validateFindUsers_request(value: any): value is void {
    return value === undefined;
}

export function validateFindUsers_response(value: any): value is {
    name: string;
    age?: number;
}[] {
    return value instanceof Array && value.reduce((acc, curr) => acc && (!(curr === null) && typeof curr === \\"object\\" && typeof curr[\\"name\\"] === \\"string\\" && (curr[\\"age\\"] === undefined || typeof curr[\\"age\\"] === \\"number\\")), true);
}

export function validateFindUsers_genericError(value: any): value is void {
    return value === undefined;
}

export function validateGetUser_request(value: any): value is void {
    return value === undefined;
}

export function validateGetUser_paramUserId(value: any): value is string {
    return typeof value === \\"string\\";
}

export function validateGetUser_response(value: any): value is {
    name: string;
    age?: number;
} {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"name\\"] === \\"string\\" && (value[\\"age\\"] === undefined || typeof value[\\"age\\"] === \\"number\\");
}

export function validateGetUser_genericError(value: any): value is void {
    return value === undefined;
}

export function validateCreateUserRequest(value: any): value is CreateUserRequest {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"name\\"] === \\"string\\";
}

export function validateCreateUserResponse(value: any): value is CreateUserResponse {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"success\\"] === \\"boolean\\";
}

export function validateGetUserResponse(value: any): value is GetUserResponse {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"name\\"] === \\"string\\";
}"
`;

exports[`TypeScript validators generator produces valid code single-file 1`] = `
"import { CreateUserRequest, CreateUserResponse, GetUserResponse } from \\"./types\\";

export function validateCreateUser_request(value: any): value is CreateUserRequest {
    return validateCreateUserRequest(value);
}

export function validateCreateUser_headerAuthToken(value: any): value is string | void {
    return value === undefined || typeof value === \\"string\\";
}

export function validateCreateUser_response(value: any): value is CreateUserResponse {
    return validateCreateUserResponse(value);
}

export function validateCreateUser_genericError(value: any): value is void {
    return value === undefined;
}

export function validateFindUsers_request(value: any): value is void {
    return value === undefined;
}

export function validateFindUsers_response(value: any): value is {
    name: string;
    age?: number;
}[] {
    return value instanceof Array && value.reduce((acc, curr) => acc && (!(curr === null) && typeof curr === \\"object\\" && typeof curr[\\"name\\"] === \\"string\\" && (curr[\\"age\\"] === undefined || typeof curr[\\"age\\"] === \\"number\\")), true);
}

export function validateFindUsers_genericError(value: any): value is void {
    return value === undefined;
}

export function validateGetUser_request(value: any): value is void {
    return value === undefined;
}

export function validateGetUser_paramUserId(value: any): value is string {
    return typeof value === \\"string\\";
}

export function validateGetUser_response(value: any): value is {
    name: string;
    age?: number;
} {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"name\\"] === \\"string\\" && (value[\\"age\\"] === undefined || typeof value[\\"age\\"] === \\"number\\");
}

export function validateGetUser_genericError(value: any): value is void {
    return value === undefined;
}

export function validateDeleteUser_request(value: any): value is void {
    return value === undefined;
}

export function validateDeleteUser_paramUserId(value: any): value is string {
    return typeof value === \\"string\\";
}

export function validateDeleteUser_headerAuthToken(value: any): value is string {
    return typeof value === \\"string\\";
}

export function validateDeleteUser_response(value: any): value is null {
    return value === null;
}

export function validateDeleteUser_genericError(value: any): value is {
    message: string;
} {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"message\\"] === \\"string\\";
}

export function validateDeleteUser_specificErrorForbidden(value: any): value is {
    message: string;
    signedInAs: string;
} {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"message\\"] === \\"string\\" && typeof value[\\"signedInAs\\"] === \\"string\\";
}

export function validateCreateUserRequest(value: any): value is CreateUserRequest {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"name\\"] === \\"string\\";
}

export function validateCreateUserResponse(value: any): value is CreateUserResponse {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"success\\"] === \\"boolean\\";
}

export function validateGetUserResponse(value: any): value is GetUserResponse {
    return !(value === null) && typeof value === \\"object\\" && typeof value[\\"name\\"] === \\"string\\";
}"
`;
